src/main/java/com/mraof/minestuck/data/BetterCombatProvider.java

```diff
@@ -0,0 +1,205 @@
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.item.Item;
+import net.neoforged.neoforge.registries.DeferredItem;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+import static com.mraof.minestuck.item.MSItemTypes.*;
+
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+public class BetterCombatProvider implements DataProvider
+{
+	private static final String SICKLE = "bettercombat:sickle";
+	private static final String SCYTHE = "bettercombat:scythe";
+	private static final String CLAW = "bettercombat:claw";
+	private static final String PICKAXE = "bettercombat:pickaxe";
+	private static final String HAMMER = "bettercombat:hammer";
+	private static final String AXE = "bettercombat:axe";
+	private static final String HEAVY_AXE = "bettercombat:heavy_axe";
+	private static final String CHAINSAW = "minestuck:chainsaw_base";
+	private static final String MACE = "bettercombat:mace";
+	private static final String SWORD = "bettercombat:sword";
+	private static final String KATANA = "bettercombat:katana";
+	private static final String CUTLASS = "bettercombat:cutlass";
+	private static final String CLAYMORE = "bettercombat:claymore";
+	private static final String RAPIER = "bettercombat:rapier";
+	private static final String DAGGER = "bettercombat:dagger";
+	private static final String LANCE = "bettercombat:lance";
+	private static final String STAFF = "bettercombat:battlestaff";
+	private static final String CANE = "minestuck:cane_base";
+	private static final String FORK = "minestuck:fork_base";
+	private static final String WAND = "bettercombat:wand";
+	private static final String HALBERD = "bettercombat:halberd";
+	private static final String SPEAR = "bettercombat:spear";
+	private static final String TRIDENT = "bettercombat:trident";
+	
+	private final PackOutput output;
+	
+	private final Map<ResourceKey<Item>, String> weaponWithParent = new HashMap<>();
+	
+	public BetterCombatProvider(PackOutput output)
+	{
+		this.output = output;
+	}
+	
+	/**
+	 * First gets all the weapons from MSItems and adds it to a Map where the key is the registry name and the value is the data.
+	 * Then it checks for custom defined weapons to use either in addition to the existing entries or to replace
+	 */
+	@Override
+	public final CompletableFuture<?> run(CachedOutput cache)
+	{
+		List<CompletableFuture<?>> futures = new ArrayList<>();
+		Path basePath = this.output.getOutputFolder(PackOutput.Target.DATA_PACK).resolve(Minestuck.MOD_ID).resolve("weapon_attributes");
+		
+		MSItems.REGISTER.getEntries().forEach(weaponHolder ->
+		{
+			if(weaponHolder.get() instanceof WeaponItem weaponItem && weaponItem.getToolType() != null)
+			{
+				String parent = getWeaponParent(weaponItem.getToolType());
+				
+				if(!parent.isEmpty())
+					weaponWithParent.put(weaponHolder.getKey(), parent);
+			}
+		});
```

Comment:
Idea: extract this block as its own function, call it from `customWeapons()`, and rename `customWeapons()` to `addAllWeapons()`.

Reply:
Done